app/src/main/java/gifapp/jsm/myapp/app/interview/jsmtech/gifapp/MainActivity.java

```diff
@@ -0,0 +1,185 @@
+
+        // setting the list to the adapter
+        listOfGifs = new ArrayList<>();
+        gifAdapter = new GIFAdapter(MainActivity.this, listOfGifs);
+
+        // othebe: Small typos here and there.
+        // setting the adapter to the recyle view
+        mRecycleView.setAdapter(gifAdapter);
+
+        // calling the volley method to load trending gifs by default
+        makeGifJSONRequest(trendingUrl + trendingOffset);
+
+
+        // enabling loading additional gifs
+        mRecycleView.setOnScrollListener(new EndlessRecyleOnScrollListenerStaggered(sglm) {
+            @Override
+            public void onLoadMore(int current_page) {
+                // Check if URL is equal to trending url;
+                if (calledMethod == 1) { // for searching
+                    // othebe: offset logic should be encapsulated in a funtion.
+                    offset += 25;
+                    makeGifJSONRequest(searchUrl + offset);
+                } else {
+                    trendingOffset += 25; // for trending
+
+                    // othebe: No need to pass variables since they are part of object.
+                    makeGifJSONRequest(trendingUrl + trendingOffset);
+                }
+            }
+        });
+
+        // enabling search in real time
+        mSearch.addTextChangedListener(new TextWatcher() {
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+            }
+
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+                if (s.toString().length() != 0) { // something is entered in the EditText
+                    offset = 0; // every time something new is entered, the search reinitializes
+                    calledMethod = 1; // search method
+                    searchGifs(s.toString()); // calling the search method
+                } else {
+                    // shows trending GIFs by default
+                    trendingOffset = 0;
+                    calledMethod = 0;
+                    makeGifJSONRequest(trendingUrl + trendingOffset);
+                }
+            }
+
+            @Override
+            public void afterTextChanged(Editable s) {
+            }
+        });
+    }
+
+
+    // search method which in turn calls the volley method
+    public void searchGifs(String s) {
+        String searchString = s.trim().replaceAll("\\s+", "+");
+        searchUrl = "http://api.giphy.com/v1/gifs/search?q=" + searchString + "&api_key=" + API_KEY + "&offset=";
+        calledMethod = 1;
+        makeGifJSONRequest(searchUrl + 0);
+    }
+
+    // volley request for fetching the JSON data and adding it to the list of gifs
+    public void makeGifJSONRequest(String url) {
+        mProgressBar.setVisibility(View.VISIBLE);
+        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, url, null, new Response.Listener<JSONObject>() {
```

Comment:
This could use some optimization. When loading more, we want to keep the current queue. However, when a new search is made, we want to cancel current requests and clear the queue. This also applies when typing a search string, etc. C, CA, CAK, CAKE. We only want to make a search for CAKE. Ideally, we would want some sort of pause before starting the network request, but at minimum we should clear the queue.